test/e2e/cluster_upgrade.go

```diff
@@ -131,14 +130,20 @@ func ClusterUpgradeConformanceSpec(ctx context.Context, inputGetter func() Clust
 			WaitForMachinesToBeUpgraded: input.E2EConfig.GetIntervals(specName, "wait-worker-nodes"),
 		})
 
-		By("Upgrading the machinepool instances")
-		framework.UpgradeMachinePoolAndWait(ctx, framework.UpgradeMachinePoolAndWaitInput{
-			ClusterProxy:                   input.BootstrapClusterProxy,
-			Cluster:                        clusterResources.Cluster,
-			UpgradeVersion:                 input.E2EConfig.GetVariable(KubernetesVersionUpgradeTo),
-			WaitForMachinePoolToBeUpgraded: input.E2EConfig.GetIntervals(specName, "wait-machine-pool-upgrade"),
-			MachinePools:                   clusterResources.MachinePools,
-		})
+		// Only attempt to upgrade MachinePools if they were provided in the template,
+		// also adjust the expected workerMachineCount if we have MachinePools
+		var workerMachineCount = clusterTemplateWorkerMachineCount
+		if len(clusterResources.MachinePools) > 0 {
+			workerMachineCount = 2 * workerMachineCount
```

Comment:
```suggestion
			workerMachineCount += int64(len(clusterResources.MachinePools)) * workerMachineCount
```
This way it would also work with >1 MachinePools, what do you tyink?

Reply:
Added a couple of helper methods to the ApplyClusterTemplateAndWaitResult type to support both multiple MachinePools and multiple MachineDeployments. I expected there to be additional uses of this in other tests, but it looks like this was the only test that did this type of computation, so the updated refactor might be a bit overkill 😂